### PR TITLE
[ALOY-1365] Added support for Alloy-specific i18n and platform folder…

### PR DESCRIPTION
**Pull request by cb1kenobi**
https://jira.appcelerator.org/browse/ALOY-1365

[ALOY-1365] Added support for Alloy-specific i18n and platform folders for both the whole Alloy app and overridable at the theme level. Improved widget i18n support. Updated all occurrences of path.existsSync() to fs.existsSync() in the primary Alloy compiler file, but there are many others.

### CODE REVIEW

File: Alloy/commands/compile/compilerUtils.js [hunk 1]
```diff
@@ -665,60 +665,56 @@ exports.copyWidgetResources = function(resources, resourceDir, widgetId, opts) {
...
-		files = wrench.readdirSyncRecursive(srcI18nDir),
-		serializer = new XMLSerializer();
+	(function walk(src, dest) {
+		if (!fs.existsSync(src)) return;
 
-	if (!fs.existsSync(buildI18nDir)) {
-		wrench.mkdirSyncRecursive(buildI18nDir, 0755);
-		fs.existsSync(appI18nDir) && wrench.copyDirSyncRecursive(appI18nDir, buildI18nDir, {preserve: false});
-	}
+		fs.readdirSync(src).forEach(function (name) {
```
Comment (FokkeZB) [inline Alloy/commands/compile/compilerUtils.js]: After this line we need this first:

```
if (path.extname(name) !== '.xml') return;
```

Or else, if an i18n dir has a `.DS_Store` for example, Alloy will crash with:

```
			_.each(destDoc.getElementsByTagName('string'), function (node) {
			              ^

TypeError: Cannot read property 'getElementsByTagName' of null
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:700:18
    at Array.forEach (native)
    at walk (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:675:23)
    at Object.mergeI18N (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:719:3)
    at parseAlloyComponent (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:828:6)
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:361:6
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/fokkezb/dev/forks/alloy/Alloy/lib/alloy/underscore.js:79:11)
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:350:6
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/fokkezb/dev/forks/alloy/Alloy/lib/alloy/underscore.js:79:11)
    at module.exports (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:346:4)
    at Object.<anonymous> (/Users/fokkezb/dev/forks/alloy/Alloy/alloy.js:113:46)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Users/fokkezb/dev/forks/alloy/bin/alloy:3:1)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:139:18)
    at node.js:999:3
```

File: Alloy/commands/compile/compilerUtils.js [hunk 2]
```diff
@@ -665,60 +665,56 @@ exports.copyWidgetResources = function(resources, resourceDir, widgetId, opts) {
...
-						root.appendChild(node);
-						root.appendChild(outxml.createTextNode('\n'));
-					}
-				});
+			// merge!
+			var existing = {};
+			var destXml = U.XML.parseFromFile(destFile);
+			var destDoc = destXml.documentElement;
+			var srcXml = U.XML.parseFromFile(srcFile);
+			var srcDoc = srcXml.documentElement;
```
Comment (FokkeZB) [inline Alloy/commands/compile/compilerUtils.js]: After this we should test if the XML is valid:

```
			if (!destDoc) {
				U.die('Error processing "' + destFile + '"');
			}

			if (!srcDoc) {
				U.die('Error processing "' + srcFile + '"');
			}
```

File: Alloy/commands/compile/compilerUtils.js [hunk 3]
```diff
@@ -665,60 +665,56 @@ exports.copyWidgetResources = function(resources, resourceDir, widgetId, opts) {
...
-
-				_.each(outxml.getElementsByTagName('string'), function(node) {
-					var name = node.getAttribute('name');
-					obj[name] = node;
-				});
 
-				_.each(sourcexml.getElementsByTagName('string'), function(node){
-					var name = node.getAttribute('name'),
-						value = node.childNodes[0].nodeValue;
+			if (!fs.existsSync(destFile)) {
```
Comment (FokkeZB) [inline Alloy/commands/compile/compilerUtils.js]: Before this line we first need to skip non-XML files from copying and merging:

```
if (path.extname(name) !== '.xml') return;
```

Or else, if an i18n dir has a `.DS_Store` for example, Alloy will crash with:

```
			_.each(destDoc.getElementsByTagName('string'), function (node) {
			              ^

TypeError: Cannot read property 'getElementsByTagName' of null
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:700:18
    at Array.forEach (native)
    at walk (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:675:23)
    at Object.mergeI18N (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/compilerUtils.js:719:3)
    at parseAlloyComponent (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:828:6)
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:361:6
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/fokkezb/dev/forks/alloy/Alloy/lib/alloy/underscore.js:79:11)
    at /Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:350:6
    at Array.forEach (native)
    at Function._.each._.forEach (/Users/fokkezb/dev/forks/alloy/Alloy/lib/alloy/underscore.js:79:11)
    at module.exports (/Users/fokkezb/dev/forks/alloy/Alloy/commands/compile/index.js:346:4)
    at Object.<anonymous> (/Users/fokkezb/dev/forks/alloy/Alloy/alloy.js:113:46)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Users/fokkezb/dev/forks/alloy/bin/alloy:3:1)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:139:18)
    at node.js:999:3
```

File: Alloy/commands/compile/index.js [hunk 4]
```diff
@@ -248,6 +253,59 @@ module.exports = function(args, program) {
 			}
 		});
 	}
+
+	// copy the platform and theme platform directories
```
Comment (FokkeZB) [inline Alloy/commands/compile/index.js]: We **must** and **can** make this change non-breaking. If we don't, people will loose their platform and i18n files at project-level.

Simply after this line add:

```
	if (fs.existsSync(path.join(paths.project, 'platform')) && !fs.existsSync(path.join(paths.app, 'platform'))) {
		U.die('Please move the /platform folder to /app/platform for Alloy 1.8.0 or later.');
	}
```

File: Alloy/commands/compile/index.js [hunk 5]
```diff
@@ -248,6 +253,59 @@ module.exports = function(args, program) {
...
+	// init i18n search paths
+	var sourceI18NPaths = [ path.join(paths.app, 'i18n') ];
+	if (theme) {
+		sourceI18NPaths.push(path.join(paths.app, 'themes', theme, 'i18n'));
+	}
+	// widgetDirs.forEach(function (widget) {
+	// 	sourceI18NPaths.push(path.join(widget.dir, 'i18n'));
+	// });
+	var destI18NDir = path.join(paths.project, 'i18n');
+	if (fs.existsSync(destI18NDir)) {
```
Comment (FokkeZB) [inline Alloy/commands/compile/index.js]: Same here:

```
		if (!fs.existsSync(sourceI18NPaths[0])) {
			U.die('Please move the /i18n folder to /app/i18n for Alloy 1.8.0 or later.');
		}
```

File: Alloy/commands/compile/compilerUtils.js (hunk 2 above)
Comment (cb1kenobi) [inline Alloy/commands/compile/compilerUtils.js]: Fixed.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,12 @@
 ## Alloy Release Notes
 
+### Release 1.8.0
```
Comment (FokkeZB) [inline CHANGELOG.md]: We haven't been using `CHANGELOG.md` anymore between 1.7.7 en 1.7.35 and @jawa9000 is in the process of moving them to the [docs](http://docs.appcelerator.com/platform/latest/#!/guide/Alloy_Release_Notes).

It's fine, will delete it later.

Comment (cb1kenobi) [inline CHANGELOG.md]: That's fine. I still update the changelog in all my other projects that have a changelog.

**Comment by FokkeZB** [comment]
**AAAAAAAPROVED**

Gone are the 👞👞

File: Alloy/commands/new/index.js [hunk 7]
```diff
@@ -72,7 +80,7 @@ module.exports = function(args, program) {
 	);
```
Comment (FokkeZB) [inline Alloy/commands/new/index.js]: Just above this we're copying a `.gitignore` that needs to have `i18n` and `platform` added as well. The file copied is at https://github.com/cb1kenobi/alloy/blob/fe48f9a4b877e5933a216cc2cce72963cf1c6274/Alloy/template/gitignore.txt

File: Alloy/commands/compile/index.js [hunk 8]
```diff
@@ -96,6 +98,50 @@ module.exports = function(args, program) {
...
+			} else if (/^\/?platform(\/.*)?$/.test(line)) {
+				folders['/platform'] = true;
+			} else if (/^\/?Resources(\/.*)?$/.test(line)) {
+				folders['/Resources'] = true;
+			}
+		});
+		var warned = false;
+		Object.keys(folders).some(function (dir) {
+			if (!folders[dir]) {
+				logger.warn('Generated "' + dir + '" folder is not in ignored by Git, please add it to your .gitignore');
```
Comment (FokkeZB) [inline Alloy/commands/compile/index.js]: 30 lines up it's *directory*, here it's *folder*. What will it be?

Comment (cb1kenobi) [inline Alloy/commands/compile/index.js]: Fixed. https://github.com/appcelerator/alloy/pull/774/commits/21f2022243a72c4c1150dd2aa4cde0448c7107fe

**Pull request merged**